File: backend/signals/latency_tracker.py
"""This module tracks response timing so answer latency can be turned into a bounded load signal without introducing any extra external dependencies."""

from __future__ import annotations

from collections import deque
from statistics import mean
from time import monotonic


class LatencyTracker:
    def __init__(self, max_samples: int = 20) -> None:
        self._question_started_at: dict[str, float] = {}
        self._history: dict[str, deque[float]] = {}
        self._max_samples = max_samples

    def mark_question_presented(self, session_id: str) -> None:
        self._question_started_at[session_id] = monotonic()

    def record_latency(self, session_id: str, latency_ms: int | None = None) -> dict[str, float]:
        measured_ms = latency_ms
        if measured_ms is None and session_id in self._question_started_at:
            measured_ms = int((monotonic() - self._question_started_at[session_id]) * 1000)
        measured_ms = max(0, int(measured_ms or 0))

        history = self._history.setdefault(session_id, deque(maxlen=self._max_samples))
        history.append(float(measured_ms))
        baseline = mean(history) if history else float(measured_ms)

        # Longer-than-baseline responses indicate elevated effort, capped to 100.
        raw_score = min(100.0, max(0.0, (measured_ms / max(baseline, 1.0)) * 35.0))
        return {"latency_ms": float(measured_ms), "baseline_ms": round(baseline, 2), "raw_score": round(raw_score, 2)}


latency_tracker = LatencyTracker()

File: frontend/src/engine/latencyTracker.ts
/**
 * Latency Tracker — ported from backend/signals/latency_tracker.py
 * Tracks response timing and converts answer latency into a bounded load signal.
 */

export class LatencyTracker {
  private questionStartedAt: Map<string, number> = new Map();
  private history: Map<string, number[]> = new Map();
  private maxSamples: number;

  constructor(maxSamples = 20) {
    this.maxSamples = maxSamples;
  }

  markQuestionPresented(sessionId: string): void {
    this.questionStartedAt.set(sessionId, performance.now());
  }

  recordLatency(
    sessionId: string,
    latencyMs?: number | null
  ): { latencyMs: number; baselineMs: number; rawScore: number } {
    let measuredMs = latencyMs ?? null;

    if (measuredMs === null && this.questionStartedAt.has(sessionId)) {
      measuredMs = Math.round(performance.now() - this.questionStartedAt.get(sessionId)!);
    }
    measuredMs = Math.max(0, measuredMs ?? 0);

    let hist = this.history.get(sessionId);
    if (!hist) {
      hist = [];
      this.history.set(sessionId, hist);
    }
    hist.push(measuredMs);
    if (hist.length > this.maxSamples) {
      hist.shift();
    }

    const baseline = hist.length > 0 ? hist.reduce((a, b) => a + b, 0) / hist.length : measuredMs;

    const rawScore = Math.min(100, Math.max(0, (measuredMs / Math.max(baseline, 1)) * 35));

    return {
      latencyMs: measuredMs,
      baselineMs: Math.round(baseline * 100) / 100,
      rawScore: Math.round(rawScore * 100) / 100,
    };
  }
}

/** Singleton instance shared across the app. */
export const latencyTracker = new LatencyTracker();
